server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java

```diff
@@ -128,23 +129,39 @@ public SearchPhaseController.ReducedQueryPhase reduce() throws Exception {
         pendingMerges.sortBuffer();
         final TopDocsStats topDocsStats = pendingMerges.consumeTopDocsStats();
         final List<TopDocs> topDocsList = pendingMerges.consumeTopDocs();
-        final List<InternalAggregations> aggsList = pendingMerges.consumeAggs();
+        SearchPhaseController.ReducedQueryPhase reducePhase;
         long breakerSize = pendingMerges.circuitBreakerBytes;
-        if (hasAggs) {
-            // Add an estimate of the final reduce size
-            breakerSize = pendingMerges.addEstimateAndMaybeBreak(PendingMerges.estimateRamBytesUsedForReduce(breakerSize));
-        }
-        SearchPhaseController.ReducedQueryPhase reducePhase = SearchPhaseController.reducedQueryPhase(
-            results.asList(),
-            aggsList,
-            topDocsList,
-            topDocsStats,
-            pendingMerges.numReducePhases,
-            false,
-            aggReduceContextBuilder,
-            rankCoordinatorContext,
-            performFinalReduce
-        );
+        try {
+            final List<DelayableWriteable<InternalAggregations>> aggsList = pendingMerges.consumeAggs();
+            if (hasAggs) {
+                // Add an estimate of the final reduce size
+                breakerSize = pendingMerges.addEstimateAndMaybeBreak(PendingMerges.estimateRamBytesUsedForReduce(breakerSize));
+            }
+            final List<InternalAggregations> aggregations = new AbstractList<>() {
+                @Override
+                public InternalAggregations get(int index) {
+                    return aggsList.get(index).expand();
```

Comment:
If I understand correctly, we only iterate this list once? So we're not expanding `DelayableWriteable` more than once?

Reply:
yes, I change the signature of the method to a collection because I want to know the size but yes, the idea is that it behaves like an iterator.